### PR TITLE
Added Manga18Me

### LINKED ISSUE #2598: Source request: Manga18
### Source name

Manga18

### Source link

https://manga18.me/

### Source language

English

### Other details

18+/NSFW = yes

There are other extensions with somewhat similar names (Manga18.Club, Manga18fx), but their respective icons and the daily content they publish are different.

### Acknowledgements

- [X] I have checked that the extension does not already exist by searching the [extension listing](https://keiyoushi.github.io/extensions/) and verified it does not appear there.
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a meaningful title with the source name.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by KenjieDec**
Checklist:
Closes #2598 

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

### CODE REVIEW

File: src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+    override val name = "Manga18Me"
+
+    override val lang = "en"
+
+    override val baseUrl = "https://manga18.me"
+
+    override val supportsLatest = true
+
+    override val client = network.client.newBuilder()
```
Comment (AwkwardPeak7) [inline src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt]: `cloudflareClient`

File: src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+import org.jsoup.nodes.Document
+import org.jsoup.nodes.Element
+import rx.Observable
+import java.lang.Exception
+import java.text.SimpleDateFormat
+import java.util.Locale
+
+class Manga18Me : ParsedHttpSource() {
+
+    override val name = "Manga18Me"
```
Comment (AwkwardPeak7) [inline src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt]: nitpick: name should be Manga18.me

File: src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        date_upload = try {
+            dateFormat.parse(element.selectFirst("span")!!.text())!!.time
+        } catch (_: Exception) {
+            0L
+        }
+    }
+
+    override fun pageListParse(document: Document): List<Page> {
+        val contents = document.select("div.read-content.wleft img")
+            ?: throw Exception("Unable to find script with image data")
```
Comment (AwkwardPeak7) [inline src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt]: `.select()` isn't nullable

File: src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+    private val infoElementSelector = "div.post_content"
+    private val titleSelector = "div.post-title.wleft > h1"
+    private val descriptionSelector = "div.ss-manga > p"
+    private val statusSelector = "div.post-content_item.wleft:contains(Status) div.summary-content"
+    private val altNameSelector = "div.post-content_item.wleft:contains(Alternative) div.summary-content"
+    private val genreSelector = "div.href-content.genres-content > a[href*=/manga-list/]"
+    private val authorSelector = "div.href-content.artist-content > a"
+    private val artistSelector = "div.href-content.artist-content > a"
+    private val thumbnailSelector = "div.summary_image > img"
```
Comment (AwkwardPeak7) [inline src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt]: inline these

File: src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        return if (query.isEmpty()) {
+            client.newCall(searchMangaRequest(page, query, filters))
+                .asObservableSuccess()
+                .map(::searchMangaParse)
+        } else {
+            client.newCall(GET("$baseUrl/search?q=$query", headers))
+                .asObservableSuccess()
+                .map(::searchMangaParse)
+        }
+    }
```
Comment (AwkwardPeak7) [inline src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt]: no need to override the fetch method, just add condition to `searchMangaRequest`

also, use `HttpUrl.Builder` to ensure query is properly encoded

File: src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+                if (searchText.lowercase().contains("raw")) {
+                    true
+                } else if (raw.lowercase().contains("raw")) {
+                    true
+                } else if (title.contains("raw")) {
+                    false
+                } else {
+                    true
+                }
+            }
```
Comment (AwkwardPeak7) [inline src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt]: what is this for?

**Review by AwkwardPeak7 — CHANGES_REQUESTED** [review]

File: src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt (hunk 6 above)
Comment (KenjieDec) [reply src/en/manga18me/src/eu/kanade/tachiyomi/extension/en/manga18me/Manga18Me.kt]: Hello! Sorry for the messy code! It seems that the website is mixing both the Raw and English comics in their latest updates. I noticed that there's a separate section for the Raw comics on their website, so I wrote that filter to exclude the Raw comics unless they are specifically searched for.

Is it not allowed to do that? Thank you

**Review by AwkwardPeak7 — APPROVED** [review]

**Pull request merged**